### DISCUSSION TITLE
What is the better approach for a multi-agent system?

### DISCUSSION THREAD

**Discussion by Arquinz** [langroid/langroid#546] [General]
Hello, I got started on Langroid yesterday and I'm loving the ability to run local models in distributed agents. I am currently in the process of building a prototype for an intelligent assistant for my University's aiot space (just a demo to begin first) using LLama 3.1-8b-Instruct.

I'd like to know what would a good approach be for a multi-agent system. My idea was to have a two-step process.

1. Interface agent that classifies the query (workspace, university or other)
2. Based on that send it to an agent which has access to either relevant data (like inventory list in csv), an agent with access to general uni documents and info and a chat agent which acts as a normal LLM chatbot.

However, I'm not sure if it would be better to manually make the interface agent classify the response, then use if-else conditions to parse the response category and initialize the relevant agent, or if it is better to divide these into tasks and make a three agent loop. I think the benefit of that is that non-relevant questions later in the conversation can automatically be routed to the right agent with DO-NOT-KNOW.

If you also have any other tips, feel free to share them. I'm really looking forward to creating the first real personal AI assistant and i'm happy langroid has given the tools to do so easily.

**Comment by pchalasani** [comment]
(EDITED)

@Arquinz glad you like the library. Since the interface agent is nothing more than a triage agent, your first approach may be better, i.e have it classify the incoming question using a tool that you define, e.g. `TypeTool`, which has a field `question_type` that your agent responder will recognize and send to the appropriate agent. 

So you'd set up a `triage_agent`, enable it to use the `TypeTool`, and wrap it in a `triage_task`. Its sole purpose is to return a `TypeTool` for unambiguous, deterministic handling. Also set up the agents + tasks for the specialized agents. You can then have a loop like this:

```python
while True:
    question = # get question from user
    result = triage_task.run(question)
    tool = result.tool_messages[0]
    q_type = tool.question_type
    # your if/then conditions to route question to one of the specialized tasks
```

There are numerous examples and tests you can see for inspiration. To see how an agent can return a tool as a result, search for `AgentDoneTool` , `DoneTool`, and how it's used in the examples.

I'm not sure what you meant by this -- 

> I think the benefit of that is that non-relevant questions later in the conversation can automatically be routed to the right agent with DO-NOT-KNOW.
